### PR TITLE
Align detached data format with attached pending data (prototype)

### PR DESCRIPTION
**Pull request by dannimad**
## Description
This is a prototype to change current serialize and rehydrateDetachedContainerFromSnapshot APIs to match current format used when collecting pending state while attached. The draft PR is created not intended to merge but for show purposes.

Current serialize API returns a stringified ISummaryTree. In order to consume it, rehydration receives that summary and converts it into an ISnapshotWithBlobContents. This change can be summarized as moving that summary -> snapshot conversion at serialization time. Such a conversion was not exactly equal to current one in order to provide a format similar to IPendingContainerState (ISnapshotTree and ISerializableBlobContents instead of ISnapshotWithBlobContents). 

## TODO
1.  There are still some tests failing. Haven't found the root cause but there are issues serializing/rehidrating multiple times and it also seems like there is some information lost in the rehydrated datastore
2.  Add back compatibility

### CODE REVIEW

**Comment by dannimad** [comment]
This is a list of current failed tests and its output: 
` 1) Dehydrate Rehydrate Container Test
       Non-Compat
         Rehydrate container multiple times round trip serialize/deserialize:
     Failed to get Channel: Error: channelTypeNotAvailable: /sm1Key
  Error: <<generated stack>>
      at generateErrorWithStack (/workspaces/FluidFramework/packages/utils/telemetry-utils/src/errorLogging.ts:203:14)
      at createResponseError (/workspaces/FluidFramework/packages/runtime/runtime-utils/src/dataStoreHelpers.ts:102:45)
      at RuntimeWithRequestHandler.request (/workspaces/FluidFramework/packages/runtime/datastore/src/dataStoreRuntime.ts:413:32)
      at RuntimeWithRequestHandler.request (/workspaces/FluidFramework/packages/runtime/datastore/src/dataStoreRuntime.ts:1140:21)
      at Context.<anonymous> (file:///workspaces/FluidFramework/packages/test/test-end-to-end-tests/src/test/deRehydrateContainerTests.spec.ts:473:22)

base.js:305
Process exited with code 3
  2) Dehydrate Rehydrate Container Test
       Non-Compat
         Rehydrate container, don't load a data store and then load after container attachment. Make changes to dds from rehydrated container and check reflection of changes in other container:
     TypeError: dataStore3.getSharedObject is not a function
      at Context.<anonymous> (file:///workspaces/FluidFramework/packages/test/test-end-to-end-tests/src/test/deRehydrateContainerTests.spec.ts:710:41)

base.js:305
  3) Dehydrate Rehydrate Container Test
       Non-Compat
         Rehydrate container, create but don't load a data store. Attach rehydrated container and load container 2 from another loader. Then load the created dataStore from container 2, make changes to dds in it check reflection of changes in rehydrated container:
     TypeError: dataStore3.getSharedObject is not a function
      at Context.<anonymous> (file:///workspaces/FluidFramework/packages/test/test-end-to-end-tests/src/test/deRehydrateContainerTests.spec.ts:762:41) `

1) blobs
       Non-Compat
         serialize/rehydrate then attach:
     Error: channelTypeNotAvailable
      at loadChannelFactoryAndAttributes (/workspaces/FluidFramework/packages/runtime/datastore/src/channelContext.ts:153:9)
      at LazyPromise.execute (/workspaces/FluidFramework/packages/runtime/datastore/src/remoteChannelContext.ts:80:36)

base.js:305
  2) blobs
       Non-Compat
         serialize/rehydrate multiple times then attach:
     Error: channelTypeNotAvailable
      at loadChannelFactoryAndAttributes (/workspaces/FluidFramework/packages/runtime/datastore/src/channelContext.ts:153:9)
      at LazyPromise.execute (/workspaces/FluidFramework/packages/runtime/datastore/src/remoteChannelContext.ts:80:36)

base.js:305
  3) blobs
       Non-Compat
         rehydrating without detached blob storage results in error:
     AssertionError [ERR_ASSERTION]: Missing expected rejection.
      at Context.<anonymous> (file:///workspaces/FluidFramework/packages/test/test-end-to-end-tests/src/test/blobs.spec.ts:590:3)

File: packages/loader/container-loader/api-report/container-loader.api.md [hunk 1]
```diff
@@ -99,6 +100,16 @@ export interface IParsedUrl {
     version: string | null | undefined;
 }
 
+// @public (undocumented)
+export interface IPendingDetachedContainerState {
```
Comment (anthony-murphy) [inline packages/loader/container-loader/api-report/container-loader.api.md:104]: i wouldn't expect any new exports

File: packages/loader/container-loader/src/utils.ts [hunk 2]
```diff
@@ -154,6 +159,57 @@ function convertSummaryToSnapshotWithEmbeddedBlobContents(
 	return treeNode;
 }
 
+/**
+ * Converts summary tree (for upload) to snapshot tree and blobs.
+ * @param summary - summary to convert
+ */
+function convertSummaryToSnapshotAndBlobs(
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/utils.ts:166]: this should have unit tests

File: packages/loader/driver-utils/src/summaryForCreateNew.ts [hunk 3]
```diff
@@ -22,6 +23,28 @@ export interface CombinedAppAndProtocolSummary extends ISummaryTree {
 		[".protocol"]: ISummaryTree;
 	};
 }
+/**
+ * @internal
+ */
```
Comment (anthony-murphy) [inline packages/loader/driver-utils/src/summaryForCreateNew.ts:28]: just keep all of this inside the container loader package

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 4]
```diff
@@ -963,6 +963,7 @@ export class ContainerRuntime
 	}
 
 	public readonly options: ILoaderOptions;
+	private waitBlobsToAttach: boolean | undefined;
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/containerRuntime.ts:966]: this seems un-related. would be good it this change was isolated

**Pull request closed without merging**